vmad/lib/fastpm.py

```diff
@@ -612,17 +611,15 @@ def gravity(self, dx):
         f = linalg.stack(r1, axis=-1)
         return dict(f=f, potk=p)
 
-    @autooperator('rhok->dx,p,f')
-    def run(self, rhok):
+    @autooperator('rhok, Om0->dx,p,f')
+    def run(self, rhok, Om0):
```

Comment:
Thanks. However,it looks like Om0 is not used anywhere in run() or leapfrog.

Reply:
It multiplies the KickFactor in both run() and leapfrog() to make dp

Reply:
Could you point to me the line where this happens?

Line 502 in this version still tries to pass the Om0 symbol into MatterDominated as a floating point. I think that's going to break, did you get an exception?

Reply:
Line 456, 467, 629, 640 are where Om0 multiplies the kick factor. And you're right, MatterDominated should not take a symbol. If I change it back to a float it will work, but will it change the overall derivative wrt Om0?

Reply:
Hi Yu,  for the MatterDominated function in fastpm-python.background we want to be able to take the derivative with respect to Om0. Does it make sense to build a differentiable ODE solver?

Reply:
It does, but it can be some non-trivial work that involves quite a bit of studying, if we want to do it properly. Since there are many steps in the ode solver, in general we will prefer to solve an adjoint equation (more efficient with less Python overhead) instead of rewriting the solve as a vmad graph, run it on a tape, then do the vjp back-propagation.

I will support you if you want to invest in the adjoint equation approach.

If you do not want to do this heavy lifting, then a much easier alternative is to do a finite differentiation. This is quite efficient since we only backprop into a single input symbol Om0:

1. Compute the power spectrum for Om0,
2. Compute the power specturm for Om0 + epsilon,
3. Compute difference of the power spectrum. Delta P (it's a vector)
4. I think the back-propagation is _Om0 = <DeltaP,  _P> / epsilon.

We can probably add this as a generic operator that does finite differentiation on an input function -- so it might end up being more useful.

I'd suggest we do the finite diff first -- and if the result is not good enough, or if you have the extra bandwidth, then we will code up the adjoint ODE solver. What do you think?

Reply:
This sounds great! I will get working on the finite diff function and keep you posted for when I get that done. Regardless of the result, I would like to code the adjoint ODE solver. It sounds cool and I think it would be really good for me to learn. If you have any reading to get me ahead, send it my way. Should we remove this pull request and wait until there is something working?

Reply:
Hi Yu,
when you say power spectrum, did you mean the the growth factor D and its derivatives dD/da d^2D/da^2, because that's what gets computed by the differential equation right? All the other functions that go into the Kick- and Driftfactor seem trivially differentiable, so maybe we just need to do finite differencing of those two?
Thanks!
Vanessa

Reply:
My understanding is these were two issues that both need to be dealt with: the linear power spectrum as a function of Om0, and the Kick Drift factors. A generic finite-diff operator should be able to deal with both, I think.

Reply:
Ok cool. I wrote a generic finite-diff operator that works on an arbitrary input function and hopefully could work with the CLASS transfer function/power spec as well. 

Right now I am coding up an autooperator for kick and drift in MADLens, and autooperators for E, G, D, etc. I guess the way I'm organizing it in my head is when we call kick and drift, they call E,G,D, and they call operators _D1, _D2, which have a vjp/jvp that uses finite differencing of the ODE. I guess I have 3 questions, 

1.) Should the apl of the finite diff operator just return the output of the function? (I think this should be the case)

2.) Am I wildly overthinking/overworking this? can I just build an autooperator kick and drift which calls the finite diff operator and this will also take care of the solution to the ODE? (I think this is not the case)

3.) the finite diff operator I made takes an arbitrary function as an input, but I can't find a way to take a class method as an input. Is there a simple way to do this?

All in all, I think I have a bunch of the pieces of the puzzle built but am a little confused about putting them all together. 

Thanks for all your help!

Reply:
1) Yes -- that's the bare minimum.

2) I have this feeling that nested auto-operators are slower than simple python functions. Based on that feeling I would recommend you work on the highest level of granularity, that is to replace the Kick Drift operator with a finite-diff operator.

However, I noticed this is incompatible with how  the code is currently structured -- The KickDrift factors consumes a 'pt' object instead of OmegaM. (https://github.com/rainwoodman/vmad/blob/master/vmad/lib/fastpm.py#L457)
The reason that Kick Drift factors consumes a PT object is to avoid repeated reintegration of the ODE. It is somewhat slow. If we give that up, then we can rewrite Kick Drift to consume OmegaM, and create pt inside MatterDominated() objects. 

def KickFactor(OmegaM, ai, ac, af):
   pt = MatterDominated(OmegaM)
   return ....

And use finite_diff(lambda x, ai=ai,ac=ac,af=af: KickFactor(x, ai, ac, af), OmegaM) in the loop.

This is slow, but should at least work.

If we want to reuse the pt objects, then one way is to somehow making pt a symbol and module the adjoints and vjp (and I think that's closer to your overworking approach).

A less clean, but easier way is to use an object cache / factory:

def KickFactor(OmegaM, ai, ac, af, CosmologyFactory):
   pt = CosmologyFactory(OmegaM)

And the function CosmologyFactory will return a cached MatterDominated object after the first call. We can create the CosmologyFactory object at the beginning of the leapfrog function, such that it dies when the graph is no longer used.

3. Would this not work? 

  finite_diff(lambda x: instance.Method(x, other))

Reply:
@rainwoodman thanks for guiding us through this. I have one more thing to add: in the interpolated lightcone we have a modified Driftfactor, that alreday takes a symbol (the approximate scale factors of the particles after half a step.) Should we do finite differencing wrt a and Omega_m for this operator?

Reply:
@rainwoodman Awesome, thank you so much for helping out! I'm going to try the rewrite of kick drift to call MatterDominated and see how slow it is and also try the CosmologyFactory approach as it might save some time. I'll put them in separate branches and let you know when I have something ready to get looked at. 

Thanks again

Reply:
I've been busy today and think I was able to solve this. In vmad.fastpm, I wrote a CosmologyFactory class, which creates MatterDominated objects and stores them in a cache with a hash corresponding to the Om0 used to instantiate MatterDominated. When the factory is called to make an object with the same parameters, It will just return the MatterDominated object from the cache without recalculating. I only did this for the leapfrog function, and the KickFactor, DriftFactor functions, so they are now functions of Om0 and FactoryCache. In MADLens, I changed all of the self.KickFactor calls to finite_difference() operator calls which take in the functions for Kick and Drift. 

When running MADLens with this new setup, all is working and we're getting the same forward model as before, so I think this is good news!

Changes to vmad are in my forked repo in CosmologyFactory branch 
Changes to MADLens are in my forked repo in kick_drift branch

Reply:
@maxelee: Would you mind filing a PR from your CosmologyFactory branch? Let's get it checked into vmad.

@VMBoehm: A quick answer is that I don't know. Would it be possible to rewrite the finite-diff operator to consume several variables? (gut feeling is yes) And if so, would it possible to rewrite the existing 'a' backprop in lightcone with finite-diff?